### ISSUE TITLE
False positive undefined-variable for self-referential return annotation in a PEP 695 generic class

### ISSUE THREAD

**Issue by sam-s** [pylint-dev/pylint#11211]
### Bug description

Pylint reports `E0602: Undefined variable 'uniqdict'` when a method in a [PEP 695](https://peps.python.org/pep-0695/) generic class uses the enclosing class in its return annotation.

```python
"https://github.com/pylint-dev/pylint/issues/11211"
class uniqdict[K, V](dict[K, V]):
    def __ior__(self, other, /) -> uniqdict[K, V]:
        return self
```

### Command used

```shell
pylint --rcfile=/dev/null --disable=all --enable=undefined-variable --py-version=3.12 pylint.py
```

### Pylint output

```python
pylint.py:3:35: E0602: Undefined variable 'uniqdict' (undefined-variable)
```

### Expected behavior

no diagnostics

### Pylint version

```shell
pylint 4.0.6
astroid 4.1.2
Python 3.14.6
```

### OS / Environment

MSYS_NT-10.0-26100

### Larger real-world example
```python
import collections.abc

class uniqdict[K, V](dict[K, V]):
    """A mapping whose existing keys cannot be assigned different values."""

    def __setitem__(self, key: K, value: V) -> None:
        try:
            if (old := super().__getitem__(key)) != value:
                raise ValueError(self.__class__.__name__, key, old, value)
        except KeyError:
            super().__setitem__(key, value)

    update = collections.abc.MutableMapping.update

    def __ior__(self, other, /) -> uniqdict[K, V]:
        self.update(other)
        return self
```

**Comment by kdelay** [comment]
This looks like expected behavior for the `--py-version` that was passed rather than a false positive.

Before 3.14, annotations in a class body are evaluated eagerly, and PEP 695 does not change that: the `def` executes while the class object is still being created, so the class name is not bound yet. Running the reproducer on real interpreters:

```
$ python3.12 t.py
  File "t.py", line 3, in uniqdict
    def __ior__(self, other, /) -> uniqdict[K, V]:
                                   ^^^^^^^^
NameError: name 'uniqdict' is not defined

$ python3.13 t.py     # same NameError
$ python3.14 t.py     # runs fine, PEP 649 makes annotations lazy
```

pylint 4.0.6 / astroid 4.0.4 lines up with that:

```
--py-version=3.12  -> E0602
--py-version=3.13  -> E0602
--py-version=3.14  -> no message
no --py-version (interpreter 3.14) -> no message
```

So with the version you are actually running on (3.14.6) the default invocation is already clean; `E0602` shows up only because `--py-version=3.12` was requested, and under 3.12/3.13 semantics that code does raise `NameError` at import time.

If 3.12 compatibility is wanted, quoting the annotation (`-> "uniqdict[K, V]"`) or `from __future__ import annotations` avoids it.

**Comment by Pierre-Sassoulas** [comment]
Closing this as the expected behavior.
